### ISSUE TITLE
Two keywords module issues

### ISSUE THREAD

**Issue by Nilvus** [darktable-org/darktable#4407]
**Describe the bug**
@phweyland 
Editing keywords needs to restart darktable to have them updated.
Also, some parents do not allow to add a new keyword directly on it.

**To Reproduce**

First issue:
1. Go to 'keyword modules'
2. Right-click on a keyword
3. Click on edit keyword
4. Edit and update the name
Nothing happens. Try to right click again and click to edit keyword again. No dialog appears.

Updating is made after restarting darktable.

Second issue:
1. Right-click on a parent
2. Select 'create a keyword'
3. Default behavior is to add directly the new keyword on the selected parent. The parent path appear on the dialog window. Sometimes, darktable doesn't propose the parent path, so it create the keyword on root tree. That's need to right-click then on this new keyword then update the path.

**Expected behavior**
1. Simply update directly the updated keyword/parent
2. Fix the path problem (second issue) but quite difficult to reproduce as this issue doesn't appear on all parents.

**Platform (please complete the following information):**
 - Darktable Version: master (last compile was 29th february)
 - OS: Ubuntu 19.10

**Comment by phweyland** [comment]
Thank you for reporting this. I'll look at it asap.

**Comment by phweyland** [comment]
I cannot reproduce case 1, but 1. it is very slow if a lot of images attached, 2. I get console gtk error messages. I'll look at it.

About the second issue. There are indeed 2 different behaviors.
1. you have the tags:
aaa
aaa|bbb

if the user user right-clicks on aaa and selects create tag, the dialog proposes aaa as a parent.

2. but if you have only
aaa|bbb

if the user right-clicks  on aaa and selects create tag, the tag is created at root level (because aaa doesn't exist).

I think the best way in that case, is not to have the option "create tag" in the popup menu. Would you agree ?

**Comment by Nilvus** [comment]
> if the user right-clicks on aaa and selects create tag, the tag is created at root level (because aaa doesn't exist).
> 
> I think the best way in that case, is not to have the option "create tag" in the popup menu. Would you agree ?

I'm really surprised that you can't reproduce first issue. It's the one I reproduce each time. So maybe a problem in my database...

About issue 2, I disagree with your proposal. It would be good to have possibility to create a tag even if there's one child. I never applied the parent on images, only child.

I make nature photos and I class them by species.

For example (If I'm not wrong, you speak french so I will write a part of my tag tree):

Faune|Papillons|Nymphalidés|

Consider that at first, I have only at first the child (after Nymphalidés) "Moyen Nacré". I don't apply "Faune", "Papillons" or "Nymphalidés" on any images. I just apply child, that's a choice I made. If I want to add another child, like "Petit nacré", if I understand correctly what you wrote, I want to have the possibility to create quickly a new child, even if I have only one at first (so your 2nd case "aaa|bbb" only.

On my system, I have Faune|Oiseaux and then just after about all european species by family, so quite few child for family that also have lots of childs (each type of birds of each family). Actually, if I would add one more child on "Oiseaux", I don't have the parent proposed (so "Faune|Oiseaux").

That's the problem here. Hope I'm clear.

**Comment by phweyland** [comment]
> That's the problem here. Hope I'm clear.

Let's check if I understand.
You have the tag `Faune|Papillons|Nymphalidés|Moyen Nacré` in your db but you don't have `Faune`, `Faune|Papillons` nor `Faune|Papillons|Nymphalidés`.
Right clicking on `Nymphalidés` (in tree view) you would like to be able to add `Petit Nacré` as a child. In other word, to create the tag `Faune|Papillons|Nymphalidés|Petit Nacré`.
Correct ?

This is not the way that works today. Currently we can add a child (`Petit nacré`) only on an existing tag ( `Faune|Papillons|Nymphalidés`). 

However the above process makes sense. I'll try to add the corresponding logic.

**Comment by Nilvus** [comment]
> > That's the problem here. Hope I'm clear.
> 
> Let's check if I understand.
> You have the tag `Faune|Papillons|Nymphalidés|Moyen Nacré` in your db but you don't have `Faune`, `Faune|Papillons` nor `Faune|Papillons|Nymphalidés`.
> Right clicking on `Nymphalidés` (in tree view) you would like to be able to add `Petit Nacré` as a child. In other word, to create the tag `Faune|Papillons|Nymphalidés|Petit Nacré`.
> Correct ?
> 
> This is not the way that works today. Currently we can add a child (`Petit nacré`) only on an existing tag ( `Faune|Papillons|Nymphalidés`).
> 
> However the above process makes sense. I'll try to add the corresponding logic.

That's exactly that. That's would be great to have this possibility. So more a feature request than an issue here.

About issue 1, do you have any idea how I can see if it's about my database ? Some cleanup/tests I can made ? I have try again to edit a tag and I reproduce it again (each time): update of the name is made only after darktable restart.

**Comment by phweyland** [comment]
> I have try again to edit a tag and I reproduce it again (each time): update of the name is made only after darktable restart.

What does happen if you disable/enable the tree view ? does the change appear ?

As I have (sometimes not always) some gtk messages on the console, it would be interesting to know you have them also. Like these:
```
(darktable:26203): Gtk-CRITICAL **: 17:35:04.828: gtk_tree_model_filter_convert_iter_to_child_iter: assertion 'filter_iter->stamp == filter->priv->stamp' failed

(darktable:26203): Gtk-CRITICAL **: 17:35:04.828: gtk_tree_store_set_valist: assertion 'VALID_ITER (iter, tree_store)' failed
```

So far I don't think the issue is in your db.
BTW the edit issue you describe ... is related to an existing or a non existing tag ?

**Comment by Nilvus** [comment]
@phweyland: If I disable/enable tree view, that update the name changed. So faster than restart darkable. The edit issue is on an existing tag.

When editing a tag, I have these lines on Terminal (maybe that could help you):

```
50,965395 [sql] /home/nilvus/Projets/darktable/src/common/image.c:177, function dt_image_film_roll(): prepare "SELECT folder FROM main.film_rolls WHERE id = ?1"
50,965639 [sql] /home/nilvus/Projets/darktable/src/common/image.c:220, function dt_image_full_path(): prepare "SELECT folder || '/' || filename FROM main.images i, main.film_rolls f WHERE i.film_id = f.id and i.id = ?1"
50,966078 [sql] /home/nilvus/Projets/darktable/src/common/metadata.c:382, function dt_metadata_get(): prepare "SELECT value FROM main.meta_data WHERE id = ?1 AND key = ?2 ORDER BY value"
50,966261 [sql] /home/nilvus/Projets/darktable/src/common/metadata.c:382, function dt_metadata_get(): prepare "SELECT value FROM main.meta_data WHERE id = ?1 AND key = ?2 ORDER BY value"
50,966404 [sql] /home/nilvus/Projets/darktable/src/common/metadata.c:382, function dt_metadata_get(): prepare "SELECT value FROM main.meta_data WHERE id = ?1 AND key = ?2 ORDER BY value"
50,966553 [sql] /home/nilvus/Projets/darktable/src/common/metadata.c:382, function dt_metadata_get(): prepare "SELECT value FROM main.meta_data WHERE id = ?1 AND key = ?2 ORDER BY value"
50,966688 [sql] /home/nilvus/Projets/darktable/src/common/metadata.c:382, function dt_metadata_get(): prepare "SELECT value FROM main.meta_data WHERE id = ?1 AND key = ?2 ORDER BY value"
50,966830 [sql] /home/nilvus/Projets/darktable/src/common/metadata.c:382, function dt_metadata_get(): prepare "SELECT value FROM main.meta_data WHERE id = ?1 AND key = ?2 ORDER BY value"
50,967007 [sql] /home/nilvus/Projets/darktable/src/common/tags.c:631, function dt_set_darktable_tags(): prepare "SELECT COUNT(*) FROM memory.darktable_tags"
50,967107 [sql] /home/nilvus/Projets/darktable/src/common/tags.c:663, function dt_tag_get_attached(): prepare "SELECT DISTINCT T.id, T.name, T.flags, T.synonyms, 1 AS inb FROM main.tagged_images AS I JOIN data.tags T on T.id = I.tagid WHERE I.imgid = 20052 AND T.id NOT IN memory.darktable_tags ORDER BY T.name"
50,967270 [sql] /home/nilvus/Projets/darktable/src/common/tags.c:1221, function dt_selected_images_count(): prepare "SELECT count(*) FROM main.selected_images"
50,967492 [sql] /home/nilvus/Projets/darktable/src/libs/metadata.c:153, function update(): prepare "SELECT key, value, COUNT(id) AS ct FROM main.meta_data WHERE id = ?1 GROUP BY key, value ORDER BY value"
50,970440 [sql] /home/nilvus/Projets/darktable/src/common/tags.c:631, function dt_set_darktable_tags(): prepare "SELECT COUNT(*) FROM memory.darktable_tags"
50,970558 [sql] /home/nilvus/Projets/darktable/src/common/tags.c:663, function dt_tag_get_attached(): prepare "SELECT DISTINCT T.id, T.name, T.flags, T.synonyms, 1 AS inb FROM main.tagged_images AS I JOIN data.tags T on T.id = I.tagid WHERE I.imgid = 20052 AND T.id NOT IN memory.darktable_tags ORDER BY T.name"
50,970722 [sql] /home/nilvus/Projets/darktable/src/common/tags.c:1221, function dt_selected_images_count(): prepare "SELECT count(*) FROM main.selected_images"
50,976784 [sql] /home/nilvus/Projets/darktable/src/common/history.c:1433, function dt_history_hash_get_status(): prepare "SELECT CASE  WHEN basic_hash == current_hash THEN 1  WHEN auto_hash == current_hash THEN 2  WHEN (basic_hash IS NULL OR current_hash != basic_hash) AND       (auto_hash IS NULL OR current_hash != auto_hash) THEN 4  ELSE 1 END AS status FROM main.history_hash WHERE imgid = 20052"
```

**Comment by phweyland** [comment]
> The edit issue is on an existing tag.

:(, I have it on a non existing tag ...



> When editing a tag, I have these lines on Terminal (maybe that could help you):

Thanks, I'll look at it. But no gtk error then.

**Comment by GrahamByrnes** [comment]
A naive question: when there is exactly the same line repeated (in particular for a db query) as in lines 3 - 8 of Nilvus's log, is this: 
a) a normal matter of a job needing several tries to succeed, or 
b) different parts of the job which aren't distingued in the log, or
c) a problem?

**Comment by phweyland** [comment]
> A naive question: when there is exactly the same line repeated (in particular for a db query) as in lines 3 - 8 of Nilvus's log, is this:
> a) a normal matter of a job needing several tries to succeed, or
> b) different parts of the job which aren't distingued in the log, or
> c) a problem?

These sql queries seem to be repeated but we don't see the parameters. Currently we have 6 different metadata, so I guess that something is requesting the values of these 6 metadata.
So I would say that's OK. 
However ORDER BY is less ok, as useless because this query should return only one record.

**Comment by Nilvus** [comment]
Fixed by #4415
